### ISSUE TITLE
False positive for `reportUnnecessaryComparison` with a callable

### ISSUE THREAD

**Issue by adamtheturtle** [microsoft/pyright#9932]
**Describe the bug**

`pyright` errors with "Condition will always evaluate to False" for a condition which evaluates to `True`.
In the real code that I have narrowed this down from, of course the condition is sometimes `True` and sometimes `False`.

**Code or Screenshots**

```python
from collections.abc import Callable


class A:
    def __call__(self) -> None:
        pass


class B:
    def __init__(self, c: Callable[[], None]) -> None:
        self.c = c


a = A()
b = B(c=a)
if b.c == a:  # pyright shows `reportUnnecessaryComparison` "Condition will always evaluate to False"
    print("Here")  # "Here" is printed
```

**VS Code extension or command-line**

Both VSCode and command line.
Command line is version 1.1.394.

**Comment by erictraut** [comment]
The logic for the `reportUnnecessaryComparison` check is based on some assumptions that can fail. This is an example. The problem is that comparisons in Python are implemented by calling an `__eq__` or `__ne__` method, and those methods are overridable. That means they can be implemented in any manner.

I'll see if I can adjust the heuristics to improve the behavior in this particular case. I'm not going to make any promises at this point. I might end up closing the issue as "won't fix" because the `reportUnnecessaryComparison` check will always be imprecise (i.e. it will produce false positives and false negatives in certain cases). If you find that it is not providing you enough utility to overcome its limitations, you can disable the check for your project. Or you can use a `# pyright: ignore` in specific cases where it makes a bad assumption.

**Comment by adamtheturtle** [comment]
Thank you @erictraut. In my use case I actually was using `is` rather than `==` so rather than using `# pyright: ignore` I instead am checking if `id(b.c) == id(a)` and that's a suitable workaround for me.

**Comment by erictraut** [comment]
After investigating deeper, I think that changing the heuristic here will do more damage than good. That is, I think I could make the check do the right thing in your example above, but it would result in more false positives and false negatives in other (more common) use cases. So I don't think this would be a good tradeoff. For that reason, I'm going to close this as "won't fix".

I'm glad to hear that you found a good workaround.

**Comment by adamtheturtle** [comment]
Understood and thank you for looking into this.
